### PR TITLE
Remove custom context package and use golang's only for api/server

### PR DESCRIPTION
**Pull request by tiborvass**
Although having a request ID available throughout the codebase is very
valuable, the impact of requiring a Context as an argument to every
function in the codepath of an API request, is too significant and was
not properly understood at the time of the review.

Also, mixing API-layer code with non-API-layer code makes the
latter usable only by API-layer code (one that has a notion of Context).

EDIT: This does not mean we don't want the ability to log request IDs; just that the solution was bringing more problems.

PS: the following files had simple conflicts but make sure you review them properly:
- api/server/container.go
- builder/internals.go
- daemon/container_unix.go
- daemon/create.go

Reverts #16228 and #16567

### CODE REVIEW

**Comment by tiborvass** [comment]
Ping @docker/core-maintainers @stevvooe

**Comment by duglin** [comment]
"go" doesn't really offer much of an alternative  - like thread local data might.  So is this basically saying that we don't want to do things like have a requestID on events so that people can correlate them? For example, to see that a series of container/image actions are all related to the same build cmd.

**Comment by tiborvass** [comment]
@duglin no this is not saying we don't want that. What it's saying is that the change brings more problems than it solves, and we did a bad job at reviewing.

We still want the ability to log request IDs and correlate things, but we need to agree how to do that. That means agreeing we want to use Context, and if we do, agreeing which package, and where to call it from.

There are other weird alternatives too like Goroutine Local Storage. Unless a lot of people agree it's better to carry around Context literally everywhere (down to the execdrivers), just because we want to log from everywhere, I think it's wiser to revert and think of a better solution.

**Comment by jfrazelle** [comment]
LGTM

**Comment by tiborvass** [comment]
Something's wrong with the commits referenced in the commit messages and some context is still in the code. Will reopen.

**Comment by tiborvass** [comment]
Actually this is all good; seeing the face of people who merged confused me.

Also the context package is used only in the `api/server/` folder where it is used to get API version and Request ID, which I think is fine and less controversial. An improvement could be to not have our own context package but overall the code within that folder is better with context than without. See #15953 for more.

**Comment by duglin** [comment]
If we're not planning on using the context for anything (aside from the version string) then we should just remove it all and handle the version string the way we used to.

But, before we move quickly to revert it I'd like to understand our next steps or replacement proposal. I think hastily removing is just as bad as hastily adding it in the first place.

To get the ball rolling, @tiborvass can you suggest an alternative solution that we should consider?

**Comment by calavera** [comment]
We want to maintain the context in the HTTPAPIFunc signature because it makes the api more compatible with other components, like Swarm.

**Comment by LK4D4** [comment]
As alternative solution I can propose some stuff like eventPrefix, which will allow you to combine different events together, at least it won't spoil code so much. Anyway I think we have some time to decide, because actually I didn't see many requests for such feature. And code looks extremely bad with ctx changes(I'm talking about spoiling all codebase with ctx argument, not about using context overall) now and that is destructive for opensource projects like docker, which relies on contributors so much.

**Comment by duglin** [comment]
@LK4D4 would that allow us to correlate it from the http request? It would be nice to be able to know which exact http request was related to which events.

**Comment by LK4D4** [comment]
@duglin Could you explain your use-cases then apart from build-sequence? Because "would nice to know" doesn't sound for me like real use-case for feature.

**Comment by LK4D4** [comment]
btw +1 on design of this PR

**Comment by duglin** [comment]
@LK4D4 for auditing/tracing/debugging purposes - to be able to trace actions taken on a container/image to incoming http requests, which are also tied to users. Likewise, to be able to know which actions were taken by the system itself instead of as a result of an http request.

**Comment by LK4D4** [comment]
@duglin In current implementation `RequestID` is just random string generated inside docker, so auditing and users out of question for now.
However I think this could be solved by refactoring daemon methods so it'll be possible to spawn all events in top-calls of API(which makes sense - separation of events subsystem from daemon). It's not easy task, but current implementation is unacceptable. I personally not going to tolerate such massive code spoiling for such feature(it isn't nearly critical one, sorry). Other maintainers can disagree with me.

**Comment by duglin** [comment]
Not sure why you think the current stuff doesn't work because of the random ID. The ID itself doesn't matter as long as its unique, random or not

File: api/server/server.go [hunk 1]
```diff
@@ -10,16 +10,14 @@ import (
...
 	"github.com/Sirupsen/logrus"
 	"github.com/docker/distribution/registry/api/errcode"
 	"github.com/docker/docker/api"
-	"github.com/docker/docker/context"
 	"github.com/docker/docker/daemon"
 	"github.com/docker/docker/pkg/sockets"
-	"github.com/docker/docker/pkg/stringid"
 	"github.com/docker/docker/utils"
+	"github.com/gorilla/mux"
+	"golang.org/x/net/context"
```
Comment (calavera) [inline api/server/server.go]: I think this is the wrong import

**Comment by calavera** [comment]
LGTM

**Comment by LK4D4** [comment]
LGTM

**Comment by duglin** [comment]
LGTM

**Pull request merged**